metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,234 @@
+#include "dbg_sshash.hpp"
+#include <dictionary.hpp>
+
+
+namespace mtg {
+namespace graph {
+DBGSSHash::~DBGSSHash() {}
+
+DBGSSHash::DBGSSHash(size_t k):k_(k) {
+    dict_ = std::make_unique<sshash::dictionary>();
+}
+
+DBGSSHash::DBGSSHash(std::string const& input_filename, size_t k):k_(k){
+    sshash::build_configuration build_config;
+    build_config.k = k;//
+    // quick fix for value of m... k/2 but odd
+    build_config.m = (k_+1)/2;
+    if(build_config.m % 2 == 0) build_config.m++;
+    dict_ = std::make_unique<sshash::dictionary>();
+    dict_->build(input_filename, build_config);
+}
+    std::string DBGSSHash::file_extension() const  { return kExtension; }
+    size_t DBGSSHash::get_k() const  { return k_; }
+    DeBruijnGraph::Mode DBGSSHash::get_mode() const  { return BASIC; }
+
+void DBGSSHash::add_sequence(std::string_view sequence,
+                             const std::function<void(node_index)> &on_insertion) {
+    // TODO: throw exception? :)
+        throw std::runtime_error("adding sequences not implemented");
+
+}
+
+void DBGSSHash::map_to_nodes(std::string_view sequence,
+                             const std::function<void(node_index)> &callback,
+                             const std::function<bool()> &terminate) const {
+    map_to_nodes_sequentially(sequence, callback, terminate);
+}
+
+void DBGSSHash ::map_to_nodes_sequentially(std::string_view sequence,
```

Comment:
Depending on the graph type, it may be much faster to get the successor of a known k-mer than to look for each k-mer independently.

For SSHash, this is the case. Rather than encoding the k-1 redundant characters between subsequent k-mers, we can encode the first one and keep updating the hash as we move along the sequence.